### PR TITLE
Allow consult-flymake to work across all buffers in a project

### PR DESCRIPTION
**Pull request by mohkale**
Now when `consult-flymake` is given a prefix argument it will source diagnostics from all buffers in the current project instead of simply the current buffer. When not in a project then consult-flymake will fallback to diagnostics just from the current buffer.

### CODE REVIEW

File: consult-flymake.el [hunk 1]
```diff
@@ -75,17 +80,27 @@
...
      (sort diags
-           (pcase-lambda (`(_ _ ,t1 _ ,m1 _) `(_ _ ,t2 _ ,m2 _))
+           (pcase-lambda (`(,b1 _ ,t1 _ ,m1 _) `(,b2 _ ,t2 _ ,m2 _))
              (let ((s1 (flymake--severity t1))
                    (s2 (flymake--severity t2)))
-               (or (> s1 s2) (and (= s1 s2) (< m1 m2)))))))))
+               (or (> s1 s2)
+                   (string-lessp b1 b2)
+                   (and (= s1 s2)
+                        (< m1 m2)))))))))
```
[the remark this answers is not in the document]
Comment (mohkale) [reply consult-flymake.el:89]: Update: Never mind. This doesn't work. :cry:. I'll test a few other combinations and see what works.

File: consult-flymake.el [hunk 2]
```diff
@@ -27,16 +27,21 @@
 
 (require 'consult)
 (require 'flymake)
+(require 'project)
```
Comment (minad) [inline consult-flymake.el:30]: Shouldn't be necessary? flymake requires project?

File: consult-flymake.el (hunk 1 above)
[the remark this answers is not in the document]
Comment (mohkale) [reply consult-flymake.el]: Couldn't figure out how to make it work in a single sort, but works with 2 sorts reasonably well.

File: consult-flymake.el [hunk 3]
```diff
@@ -74,18 +69,36 @@
...
-               (or (> s1 s2) (and (= s1 s2) (< m1 m2)))))))))
+     ;; Sort first by position and then by the name of the buffer.
+     (sort
+      (sort diags
+            (pcase-lambda (`(,b1 _ ,t1 _ ,m1 _) `(,b2 _ ,t2 _ ,m2 _))
+              (let ((s1 (flymake--severity t1))
+                    (s2 (flymake--severity t2)))
+                (or (> s1 s2) (and (= s1 s2) (< m1 m2))))))
+      (pcase-lambda (`(,b1 _ _ _ _ _) `(,b2 _ _ _ _ _))
+        (string-lessp b1 b2))))))
```
Comment (minad) [inline consult-flymake.el:80]: I am not fond of sorting twice. Please change the predicate instead.

File: consult-flymake.el (hunk 1 above)
[the remark this answers is not in the document]
Comment (minad) [reply consult-flymake.el]: :thinking:

[the remark this answers is not in the document]
Comment (minad) [reply consult-flymake.el]: Just fyi, `(or (> s1 s2) (and (= s1 s2) (< m1 m2))))` is the criterion which sorts first by s and then by m....

File: consult-flymake.el (hunk 3 above)
Comment (minad) [reply consult-flymake.el:80]: See my comment https://github.com/minad/consult/pull/589#discussion_r889033252. It is an exercise for you. I insist on this :)

Comment (mohkale) [reply consult-flymake.el]: Okay, got it in one try :tada:, ISTG I actually tried this out but then some entries with a given mark were sorted before others even though order wise they shouldn't have been. I'm guessing I just misread it (I was testing with buffer names that were too similair to each other so they were easy to mistake).

**Comment by mohkale** [comment]
Working like a charm :-).

Screenshot_20220603_161849

File: consult-flymake.el [hunk 4]
```diff
@@ -70,14 +70,16 @@ RAW-DIAGS should be a list of diagnostics as returned from `flymake-diagnostics'
...
+     (sort diags
+           (pcase-lambda (`(,b1 _ ,t1 _ ,m1 _) `(,b2 _ ,t2 _ ,m2 _))
+             (let ((s1 (flymake--severity t1))
+                   (s2 (flymake--severity t2)))
+               (or
+                (> s1 s2)
+                (string-greaterp b1 b2)
+                (and (= s1 s2)
+                     (string-lessp b1 b2)
+                     (< m1 m2)))))))))
```
Comment (minad) [inline consult-flymake.el:82]: This doesn't look correct.

Comment (minad) [reply consult-flymake.el:82]: What do you want to achieve?

Sort first by buffer name, then severity, then position?

Comment (minad) [reply consult-flymake.el:82]: This is still wrong. I will check tomorrow. Please think about this more thoroughly.

**Comment by mohkale** [comment]
> Btw, @mohkale what about the FSF copyright assignment? Do you have that?

Yep :-).

**Comment by minad** [comment]
Thanks!

**Comment by mohkale** [comment]
Oh, glad I got it right :sweat_smile:. That took a few attempts but nice to learn something new. I'm still not used to boolean based sorting functions but that was a nice challenge. Thanks :-).

**Pull request merged**